pm-states: Add power management state in DTS

ceolin opened the pull request:
- Add information about supported power management in device tree
- Add a couple of macros to extract this information

Commit:
dts: power: Add power states support

Add power state property binding and include this property in cpu.
The pm-state attribute is a enum that matches with enum pm_state right
now the only timing attribute is the minimum residency that is the
minimum time for a power state be worthwhile.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
dts: Add _OR variant macro for DT_PROP_BY_PHANDLE_IDX

Like DT_PROP_BY_PHANDLE_IDX(), but with a fallback to
default_value.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Add macros to get pm states from dts

Add a serie of helpers to extract power management states from device
tree. DT_PM_STATE_ITEMS_LIST generates a list of a pm_state and
DT_PM_STATE_INFO_ITEMS_LIST a list of pm_state_info.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: power: Add pm state API test

Test macros used to extract pm_state information from device tree.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

erwango commented on dts/bindings/power/state.yaml:
```diff
@@ -0,0 +1,26 @@
+        type: int
+        required: true
+        description: indicates a power state
+        enum:
+            - 0 # PM_STATE_ACTIVE
+            - 1 # PM_STATE_RUNTIME_IDLE
+            - 2 # PM_STATE_SUSPEND_TO_IDLE
+            - 3 # PM_STATE_STANDBY
+            - 4 # PM_STATE_SUSPEND_TO_RAM
+            - 5 # PM_STATE_SUSPEND_TO_DISK
+            - 6 # PM_STATE_SOFT_OFF
```
enum could use strings, which would be clearer for the reader:
```suggestion
        enum:
            - active
            - runtime_idle
            - suspend_to_idle
            ...
```

pabigot replied:
I'm inclined to agree these should be symbolic.  We need to kickstart #21273.

erwango replied:
This actually works right now already (https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/pinctrl/st%2Cstm32-pinctrl.yaml#L76).
Would we need something more for this case?

pabigot replied:
We need to make sure that the values correspond to the constants defined in the C header that are used in runtime API.  I believe the devicetree ordinals are based on the order of the text in the enum sequence, so if one or the other changes value/order things will go wrong in very subtle ways.

ceolin replied:
yep, we can if the values are previously defined right ? we can't just set strings here, right ?

ceolin replied:
that is the problem I brought in the last meeting. I agree with @erwango strings are clearer, but I really hate having this information duplicated in C domain and here. This is very fragile ... The lucky here is that this should not change very often.

Will change to use string while we don't have a better option.

mniestroj replied:
I agree that using strings looks better.

Perhaps we should generate DT macros automatically based on enum strings as a new feature in DT macros generator, like:
```
#define DT_COMPAT_PM_STATE_ACTIVE    0
#define DT_COMPAT_PM_STATE_RUNTIME_IDLE    1
...
```
then access them with something like:
```
DT_COMPAT_ENUM_IDX(pm_state, active) -> 0
DT_COMPAT_ENUM_IDX(pm_state, runtime_idle) -> 1
```

pabigot commented on include/power/power_state.h:
```diff
@@ -103,6 +103,151 @@ enum pm_state {
+ */
+struct pm_state_info {
+	enum pm_state state;
+	int32_t min_residency;
```
Make this unsigned.  We're never going to have a valid negative residency, right?  And what does a zero value mean?

ceolin replied:
-1 is option not set, 0 I don't have a concrete use case, but *may* be possible that in some platform the residency time for a specific state is not relevant than 0 can represent it.

pabigot commented on include/power/power_state.h:
```diff
@@ -103,6 +103,151 @@ enum pm_state {
+};
+
+/**
+ * @brief Construct a pm_state_info from 'pm_states' property at index 'i'
```
All this belongs in `include/devicetree/pm_state.h`.

pabigot commented on dts/bindings/power/state.yaml:
```diff
@@ -0,0 +1,26 @@
+            - 4 # PM_STATE_SUSPEND_TO_RAM
+            - 5 # PM_STATE_SUSPEND_TO_DISK
+            - 6 # PM_STATE_SOFT_OFF
+    min-residency:
```
Should be `min-residency-us` following Linux.

pabigot commented on include/power/power_state.h:
```diff
@@ -103,6 +103,151 @@ enum pm_state {
+ *		cpu0: cpu@0 {
+ *			device_type = "cpu";
+ *			...
+ *			pm-states = <&state0 &state1>;
```
This is structurally very different from [Linux](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/devicetree/bindings/arm/idle-states.yaml#n479).  Does it need to be that different?

ceolin replied:
```
  - |

            CPU_SLEEP_1_0: cpu-sleep-1-0 {
                compatible = "arm,idle-state";
                local-timer-stop;
                arm,psci-suspend-param = <0x0010000>;
                entry-latency-us = <70>;
                exit-latency-us = <100>;
                min-residency-us = <300>;
                wakeup-latency-us = <150>;
            };
...
    cpus {
        #size-cells = <0>;
        #address-cells = <1>;

        cpu@0 {
            device_type = "cpu";
            compatible = "arm,cortex-a15";
            reg = <0x0>;
            cpu-idle-states = <&cpu_sleep_1_0>;
        };
```

What exactly is that different ? Linux has more properties, some of them we will probably add later, I just didn't put now because we don't have use for then currently and it is always easy adding something than remove. Regarding names, I saw more power/pm than idle in our code.

But IMHO the struct is the same, we have power state declaration and and cpu has a property with a list of them ....

pabigot replied:
OK, I see it's not as structurally different as I thought, it's mostly that the property and compatible names don't match.

We have generally tried to be compatible at devicetree source unless there's some clear reason we can't do that.  I'm not convinced we can as Linux has target-specific compatibles and perhaps we're aiming to be target-agnostic, but I do think we should match the property names that aren't ARM or other vendor specific, such as the latencies and timer effects.

ceolin replied:
I tend to agree as well. In this pr, it translate to `s/min-residency/min-residency-us/`

nashif commented on dts/bindings/power/state.yaml:
```diff
@@ -0,0 +1,26 @@
+        required: true
+        description: indicates a power state
+        enum:
+            - "PM_STATE_ACTIVE"
```
will that (PM_STATE_ACTIVE) ever be used in this context? Maybe we should only list sleep states.

ceolin replied:
The problem is keeping in sync these values with `enum pm_state`

pabigot replied:
If we use the enum token approach that's not a problem.

I don't immediately see use of full active state in the Linux devicetree state descriptions, though I can imagine it could be useful (e.g. measures of base power usage in active mode).

ceolin replied:
the enum token approach indeed solve this problem

pabigot commented on include/devicetree/pm_state.h:
```diff
@@ -0,0 +1,165 @@
+ * @param i index of pm_states prop which type is 'phandles'
+ * @return pm_state_info item from 'pm_states' property at index 'i'
+ */
+#define DT_PM_STATE_INFO_ITEM_BY_IDX(node_id, i)                    \
```
Do we expect this to work for all architectures (i.e. no variation in state description that's specific to a given platform)?

If so, then I think all this should be able to go into a C file that defines the power state data structures.  No need to have it available for all users.

If not, then it isn't going to work here anyway as it won't support variant interpretations.

ceolin replied:
I'm not sure I understood your question. My goal here is allow one easily extract this info from device tree, I expect application be able work with it in case the app has its own pm policy.

pabigot reviewed (commented):
I think we can deal with the states in a cleaner way by making progress on #21273.

Can we design things so the conversion from devicetree power state configurations to data structures is done in one place, with API to access the constructed data structure?

ceolin commented on dts/bindings/power/state.yaml:
```diff
@@ -0,0 +1,26 @@
+        required: true
+        description: indicates a power state
+        enum:
+            - "PM_STATE_ACTIVE"
```
my two cents, I don't think `_ENUM_TOKEN` solve the main problem that is having this information duplicated in both domains (C and DTS). One still need to follow a name convention and we need to declare these values in dts bindings. That is said, the thing that `ENUM_TOKEN` gives to me is that I don't need to declare the states in the correct order in the binding.

I'm totally ok in using `ENUM_TOKEN` here when it gets merged, I just don't think that it should gate this pr.

pabigot replied:
It kinda has to, unless you have another way to set this up so the enum values don't include the `PM_STATE_` prefix.  Because that should not be there, and we shouldn't merge with bindings that we intend to change.

I think we should be able to get this in within a week, there're only a couple of bikeshed color choices relevant to enum token support.

ceolin replied:
why can't it be changed later ? I mean, if the problem is changing the binding later, I can remove the `PM_STATE` prefix.
I've more code depending in these basic changes, the whole power subsystem is still using old states and getting info from Kconfig. I want to move on in these changes .... this `ENUM_TOKEN` change will be quite small and only in the macro to get it and should not impact the rest of the changes.

pabigot replied:
Yes, please remove the `PM_STATE_` prefix from the binding `pm-state` property enumeration values.  If you can make that work without enum token support in devicetree or removing the `PM_STATE_` prefix from the pm_state C enum values, that'd be fine.

mbolivar-nordic replied:
> [ @mbolivar-nordic ] would you have time to work on `_ENUM_TOKEN` or should I take it on?

As discussed, I'm on it

pabigot replied:
thank you

pabigot commented on include/devicetree/pm_state.h:
```diff
@@ -0,0 +1,165 @@
+ * @param node_id A node identifier compatible with pm-states
+ * @return macro function to construct a pm_state_info
+ */
+#define DT_PM_STATE_INFO_ITEMS_FUC(child, node_id) \
```
What the heck does `_FUC` mean?  I see this in the nuvoton npcx macros, but it makes no sense there either.

tbursztyka replied:
I guess _FUNC would have been more adequate here.

pabigot replied:
Or maybe `_ULFUNC` as it's supposed to be a function for `UTIL_LISTIFY`?

Either would be an improvement.

pabigot replied:
^^ @mbolivar @galak do you want a standard naming convention DT macros that are intended to be used in `UTIL_LISTIFY`?

pabigot reviewed (dismissed):
Documentation, type change, and probably a change to an dysphemistic and confusing macro suffix.

Otherwise I have no objections, though I would very much like to see the API we'll actually use to access the `pm_state` data structure.

nashif commented on include/power/power_state.h:
```diff
@@ -103,6 +103,14 @@ enum pm_state {
+ */
+struct pm_state_info {
+	enum pm_state state;
+	int32_t min_residency_us;
```
After some thought, I think we should add all values, when you actually do the profiling and get the data for min-residency you probably want to add the other values as well, basically entry, exit, wakeup etc., the same way Linux does it.

In current implementation we only check for min-residency, but if someone writes their own policy, they could use other data depending on the use-case. 

Do not see any overhead in adding that and supporting it right now.

ceolin replied:
Yep, I agree that we will end up having them. But currently I don't even have reference values for these things ... I am changing the current code with this and thought add unused  information now would just bring noise to that.

Can we keep it for now, and add these properties later ?

nashif replied:
not sure why you need reference values, this should be in the bindings and for now, if they are not filled you do not need to do anything, it is not like we have reference for min-residency-us either, this will need to be done based on someone actually knowing their hardware and profiling mostly. For testing we can just make up those values.

ceolin replied:
I mean, I have residency time for a target `mec15xxevb_assy6853` that I am feeding in this board dts in my changes, but these other properties have no current use, we currently don't even have a case where I'll use it, so what I'm proposing is adding these properties after the code change.

Just add these properties in the binding is no problem, but then I need to add these properties in the pm_state_info and a lot of unused fields may be confuse during these transition.

Of course, if that is a blocker I can just add it now, it is very easy to do it.

ceolin replied:
I just don't particularly like the idea of having data the usage is not 100% clear yet. It is quite easy to forget removing them later if they are not needed ...

tbursztyka commented on include/devicetree/pm_state.h:
```diff
@@ -0,0 +1,165 @@
+ * @param node_id A node identifier compatible with pm-states
+ * @return macro function to construct a pm_state enum
+ */
+#define DT_PM_STATE_ITEMS_FUC(child, node_id) \
```
s/_FUC/_FUNC

tbursztyka reviewed (approved):
approving already, other's comments need to be applied anyway

ceolin commented:
@pabigot would you mind take a look on it again ?

pabigot commented on dts/bindings/power/state.yaml:
```diff
@@ -0,0 +1,26 @@
+        type: string
+        required: true
+        description: indicates a power state
+        enum:
```
Based on https://github.com/zephyrproject-rtos/zephyr/pull/30406#discussion_r535346182 I'm expecting the tokenizing support to allow these to be changed to lower case with hyphen separators.  That does not have to be done now, but should be done as soon as the tooling supports it.

mbolivar-nordic replied:
> Based on [#30406 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/30406#discussion_r535346182) I'm expecting the tokenizing support to allow these to be changed to lower case with hyphen separators

That is indeed the case with the latest update, just pushed

pabigot reviewed (dismissed):
I think this is OK under condition of a future change, but because it's adding system-level devicetree macros I really want either @galak or @mbolivar-nordic to sign off.

mbolivar-nordic commented on dts/bindings/cpu/cpu.yaml:
```diff
@@ -10,3 +10,9 @@ properties:
+       type: phandles
+       required: false
+       description: |
+         List of power management states supported
+         by this cpu.
```
I think this will probably look weird in the [bindings index](https://docs.zephyrproject.org/latest/reference/devicetree/bindings.html); did you consider putting all this text on one line?

ceolin replied:
This particularly one I can, but what happens with big lines ? Can't we break them ?

mbolivar-nordic replied:
> This particularly one I can, but what happens with big lines ? Can't we break them ?

Yeah, for sure, you can break them, they just get formatted in a big text block if you do. Here's an example:

https://docs.zephyrproject.org/latest/reference/devicetree/bindings/ilitek%2Cili9340-spi.html#ilitek-ili9340-spi

This cpu binding looks like it will fit on one line, which may be easier to read.

mbolivar-nordic commented on tests/subsys/power/power_states_api/src/main.c:
```diff
@@ -0,0 +1,47 @@
+#include <zephyr.h>
+#include <ztest.h>
+#include <power/power.h>
+#include <devicetree/pm_state.h>
```
We don't do it  like this. The files in include/devicetree/ are private headers which we include from the bottom of devicetree.h; please keep this consistent.

ceolin replied:
good point, will change it

mbolivar-nordic commented on include/devicetree/pm_state.h:
```diff
@@ -0,0 +1,165 @@
+extern "C" {
+#endif
+
+#include <power/power_state.h>
```
Devicetree headers shouldn't need to rely on higher level system headers, so this looks like a layering inversion to me. This include should be removed. In general we need to be able to pull in devicetree.h from linker scripts, and we need all the private devicetree header files to be included in devicetree.h, so this include should be done somewhere else, by the C files that need this.

mbolivar-nordic commented on include/devicetree/pm_state.h:
```diff
@@ -0,0 +1,165 @@
+#include <power/power_state.h>
+
+/**
+ * @defgroup devicetree-pm_states Power Management states device tree macros
```
This must be reflected in an update to doc/reference/devicetree/api.rst.

mbolivar-nordic commented on include/devicetree/pm_state.h:
```diff
@@ -0,0 +1,165 @@
+ */
+
+/**
+ * @brief Construct a pm_state_info from 'pm_states' property at index 'i'
```
I think @galak should confirm,  but I don't think it's right to place structure-creating macros in here.

I think instead the power_state.h header that defines the structure should provide a helper macro to create it from a devicetree node_id and other information as needed, similarly to DEVICE_DT_DEFINE.

mbolivar-nordic commented on include/devicetree/pm_state.h:
```diff
@@ -0,0 +1,165 @@
+	DT_PM_STATE_INFO_ITEM_BY_IDX(node_id, child)
+
+/**
+ * @brief Macro function to construct a list of 'pm_state_info' items by
```
Same, this and other structure initializer macro magic should move to the header that defines the structure.

ceolin replied:
Initially I'd added these macros in the file that define the struct. Now I see two possibilities,

1) Remove devicetree/pm_states.h and move these macros to power/power_state.h and change macros namespace.
2) Remove the `#include <power/power_state.h>` from `devicetree/pm_state.h` since any macro is explicitly referencing things from that file.

The option 1 is my favorite.

mbolivar-nordic replied:
> Remove devicetree/pm_states.h and move these macros to power/power_state.h and change macros namespace.

no objections!

ceolin replied:
great, will proceed this way.

mbolivar-nordic reviewed (dismissed):
Just some organizational and docs things. Nice work @ceolin !

ceolin commented:
@mbolivar-nordic in the end moving those macros to `power_states.h` made most of the comments not applicable. Changed though the description for `pm-states` in cpu binding and change the enum values to be downcase. Would mind take another look.

mbolivar-nordic commented on include/devicetree.h:
```diff
@@ -596,6 +596,28 @@
+ * @param default_value a fallback value to expand to
+ * @return the property's value
+ */
+#define DT_PROP_BY_PHANDLE_IDX_OR(node_id, phs, idx, prop, default_value) \
```
Can you add a dedicated test case for this to tests/lib/devicetree/api? Otherwise lgtm

ceolin replied:
done

Commit:
tests: devicetree: Add DT_PROP_BY_PHANDLE_IDX_OR test

Test macro DT_PROP_BY_PHANDLE_IDX_OR. There are two tests, one when
the property exists and other when the property is not set.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

pabigot commented on dts/bindings/power/state.yaml:
```diff
@@ -0,0 +1,26 @@
+        description: indicates a power state
+        enum:
+            - "active"
+            - "runtime_idle"
```
nit-pick: lower case with hyphens.  This is more consistent enum use in Linux bindings.

mbolivar-nordic replied:
I wasn't going to ask for this, but I did think the same thing, fwiw.

ceolin replied:
done, anything else ?

pabigot reviewed (changes requested):
One change, then good.
